### PR TITLE
JAMES-3534 Implement Identity/set update (custom identities)

### PR DESCRIPTION
**Pull request by vttranlina**
Jira: https://issues.apache.org/jira/browse/JAMES-3534
resolve https://github.com/linagora/james-project/issues/4429

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 1]
```diff
@@ -662,4 +661,438 @@ trait IdentitySetContract {
...
+    val response: String = `given`
+      .body(
+        s"""{
+           |    "using": [ "urn:ietf:params:jmap:core", "urn:ietf:params:jmap:submission" ],
+           |    "methodCalls": [
+           |        [
+           |            "Identity/set",
+           |            {
+           |            		"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+           |                "update": {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:676]: indent

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 2]
```diff
@@ -662,4 +661,438 @@ trait IdentitySetContract {
...
+           |            },
+           |            "c1"
+           |        ]
+           |    ]
+           |}
+           |""".stripMargin)
+  }
+
+  @Test
+  def updateShouldModifiedIdentityEntry(): Unit = {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:717]:
```suggestion
  def updateShouldModifyIdentityEntry(): Unit = {
```

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala [hunk 3]
```diff
@@ -662,4 +661,438 @@ trait IdentitySetContract {
...
+    val response: String = `given`
+      .body(
+        s"""{
+           |    "using": [ "urn:ietf:params:jmap:core", "urn:ietf:params:jmap:submission" ],
+           |    "methodCalls": [
+           |        [
+           |            "Identity/set",
+           |            {
+           |            		"accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
+           |                "update": {
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/IdentitySetContract.scala:1005]: indent

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+object IdentitySetUpdatePerformer {
+  sealed trait UpdateResult
+
+  case class UpdateSuccess(id: IdentityId) extends UpdateResult
+
+  case class UpdateFailure(id: UnparsedIdentitySetId, exception: Throwable) extends UpdateResult {
+    def asSetError: SetError = exception match {
+      case e: IdentityUpdateParseException => e.setError
+      case e: IdentityNotFoundException => SetError.notFound(SetErrorDescription(e.getMessage))
+      case e: InvalidPropertyException => SetError.invalidPatch(SetErrorDescription(s"${e.cause}"))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:47]:
```suggestion
      case e: InvalidPropertyException => SetError.invalidPatch(SetErrorDescription(e.cause.getMessage))
```

The string interpolation is clearly not needed here.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  case class UnparsedIdentitySetId(id: Id) {
+    def serialise: String = id.value
+
+    def parse: Either[IllegalArgumentException, IdentityId] = Try(UUID.fromString(id.value))
+      .toEither
+      .left.map({
+      case e: IllegalArgumentException => e
+      case e => new IllegalArgumentException(e)
+    }).map(uuid => IdentityId(uuid))
+  }
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:81]: We already have a `UnparsedIdentityId`. Can't we reuse it instead?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            SetErrorDescription("Some unknown properties were specified"),
+            Some(Properties.toProperties(unknownProperties.toSet)))))
+        case (specifiedServerSetProperties, _) if specifiedServerSetProperties.nonEmpty =>
+          Left(IdentityUpdateParseException(SetError.invalidArguments(
+            SetErrorDescription("Some server-set properties were specified"),
+            Some(Properties.toProperties(specifiedServerSetProperties.toSet)))))
+        case _ => scala.Right(jsObject)
+      }
+  }
+}
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:116]: These properties are redundant with what is declared in https://github.com/apache/james-project/pull/752. Please reuse the properties defined there.

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  def update(request: IdentitySetRequest, mailboxSession: MailboxSession): SMono[IdentitySetUpdateResults] =
+    SFlux.fromIterable(request.update.getOrElse(Map()))
+      .concatMap {
+        case (unparsedId, json) =>
+          val either: Either[Exception, SMono[UpdateResult]] = for {
+            identityId <- unparsedId.parse
+            updateRequest <- parseRequest(json)
+          } yield {
+            update(identityId, updateRequest, mailboxSession)
+              .onErrorResume((error: Throwable) => SMono.just[UpdateResult](UpdateFailure(unparsedId, error)))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:128]:
```suggestion
              .onErrorResume(error => SMono.just[UpdateResult](UpdateFailure(unparsedId, error)))
```

?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+          }
+          either.fold((error: Exception) => SMono.just[UpdateResult](UpdateFailure(unparsedId, error)),
+            (updateRequestResult: SMono[UpdateResult]) => updateRequestResult)
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:131]:
```suggestion
          either.fold(error => SMono.just[UpdateResult](UpdateFailure(unparsedId, error)),
            updateRequestResult => updateRequestResult)
```

?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala (hunk 7 above)
Comment (vttranlina) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:128]: the new version of IDE suggests it, maybe it is more modern?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala (hunk 6 above)
Comment (vttranlina) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:116]: different properties. 
- create: accept `email` 
- update: not accept `email`

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:116]: Sure, adapt Quan work to you needs but do not duplicate so much code :-P

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala (hunk 7 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentitySetUpdatePerformer.scala:128]: Yes I prefer implicit typing in lambdas.

**Review by chibenwa — APPROVED** [review]
Looks overall good to me!

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
rebase master & squash fixup

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**